File: paintfence.py
class Solution(object):
    def numWays(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: int
        """
        if n == 0 or k == 0:
            return 0
        if n == 1: return k
        same = [ k for _ in xrange(n)]
        diff = [ k for _ in xrange(n)]
        diff[1] = k*(k-1)
        for i in xrange(2, n):
            same[i] = diff[i-1]
            diff[i] = same[i-1]*(k-1) + diff[i-1]*(k-1)
            
        return diff[-1] + same[-1]

File: paintfence.cpp
class Solution {
public:
    int numWays(int n, int k) {
        if (n == 0 or k == 0) return 0;
        if (n == 1) return k;
        vector<vector<int>> dp(n, vector<int>(2, k));
        dp[1][1] = k*(k-1);
        for (int i = 2; i < n; i++) {
            //same
            dp[i][0] = dp[i-1][1];
            //diff
            dp[i][1] = dp[i-1][1]*(k-1) + dp[i-1][0]*(k-1);
            
        }
        return dp.back()[0]+dp.back()[1];
    }
};
